[registration] Multi-threaded NDT

koide3 opened the pull request:
This PR adds a multi-threaded and SSE-optimized NDT (for reference: https://github.com/PointCloudLibrary/pcl/pull/4180 https://github.com/PointCloudLibrary/pcl/pull/4135). Taking a diff with ndt.(h|hpp) would be helpful to see the added/updated codes.

**[Added] NormalDistributionsTransformOMP**
- The following two methods are added to the original NDT
- setNumThreads() specifies the number of threads to be used
- setNeighborSearchMethod() offers a trade-off between registration stability and speed.

Benchmark result on KITTI00 (ApproxVoxelGrid: 0.25m, voxel resolution: 2.0m, transformation epsilon 1e-3):
(Average processing time per scan and relative trajectory errors are reported)
```
--- NDT ---
Time per scan    : 496.053 +- 162.218 [msec]
RTE(10, 50, 100m): 0.320342 1.564389 3.363665


--- NDT_OMP (KDTREE, 1 thread) ---
Time per scan    : 367.684 +- 126.874 [msec]
RTE(10, 50, 100m): 0.320346, 1.564428, 3.363695

--- NDT_OMP (KDTREE, 16 threads) ---
Time per scan    : 76.216 +- 39.373 [msec]
RTE(10, 50, 100m): 0.320346, 1.564428, 3.363695


--- NDT_OMP (DIRECT1, 1 thread) ---
Time per scan    : 73.426 +- 21.062 [msec]
RTE(10, 50, 100m): 0.240691, 0.931008, 2.039546

--- NDT_OMP (DIRECT1, 16 threads) ---
Time per scan    : 23.819 +- 7.633 [msec]
RTE(10, 50, 100m): 0.240691, 0.931008, 2.039546


--- NDT_OMP (DIRECT7, 1 thread) ---
Time per scan    : 287.151 +- 95.607 [msec]
RTE(10, 50, 100m): 0.216932, 0.926056, 1.977688

--- NDT_OMP (DIRECT7, 16 threads) ---
Time per scan    : 60.003 +- 32.637 [msec]
RTE(10, 50, 100m): 0.216932, 0.926056, 1.977688


--- NDT_OMP (DIRECT26, 16 threads) ---
Time per scan    : 241.036 +- 133.761 [msec]
RTE(10, 50, 100m): 13.434766, 52.471817, 81.097012

--- NDT_OMP (DIRECT27, 16 threads) ---
Time per scan    : 138.623 +- 84.487 [msec]
RTE(10, 50, 100m): 0.293359, 1.378685, 3.139081
```
It's not a very comprehensive comparison, but we can see the trend of processing speed and accuracy.

Note:
- It's faster than the original NDT even when the number of threads is set to 1 thanks to the SSE optimization.
- KDTREE shows the same result as the original implementation (very small difference is caused by SSE single-float calculation).
- DIRECT1 is very fast, but can be unreliable when the voxel resolution is small
- DIRECT7 is faster and more accurate than KDTREE. I think this should be the default method.
- DIRECT27 is worse than DIRECT7 in terms of both processing speed and accuracy in this case. But, it may have a larger convergence basin than the other methods.
- I didn't include DIRECT26 in NDT_OMP because it looks not worthy.

kunaltyagi commented:
Tagging @shrijitsingh99 for relevance with GSoC

shrijitsingh99 commented:
> Tagging @shrijitsingh99 for relevance with GSoC
Looking into this seems as if this will be similar to how integration with filters is going to take place. The only point which would need to be taken care of is the `setNeighborSearchMethod` which doesn't seem to be present the original ndt implementation. 

Is the implementation set to replace the original CPU implementation too?

kunaltyagi commented:
Should we create a separate issue to note that the non-special NDT code needs update (till the level of NDT_OMP implementation)? If (or rather when) the executors get integrated, a large update would not be needed.

shrijitsingh99 commented:
Is an update to non-special needed? Is there any reason we are not jsut removing original NDT and renaming ndt_omp to ndt besides maybe potential bugs?

kunaltyagi commented on registration/include/pcl/registration/impl/ndt_omp.hpp:
```diff
@@ -0,0 +1,810 @@
+  computeAngleDerivatives (transform);
+
+  // Update gradient and hessian for each point, line 17 in Algorithm 2 [Magnusson 2009]
+  #pragma omp parallel for num_threads(num_threads_) schedule(guided, 8)
```
Feels like only these 2 lines are the difference. Can you point to the new additions in core logic?

koide3 replied:
Major new additions are follows:

Neighbor voxel search method
- Direct voxel accessing methods are added
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L204

Thread safety
- ```point_(jacobian|hessian)_``` are made per-thread local variables for thread safety
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L195
- Summing of some variables are modified in a thread-safe manner
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L175
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L241
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L250

SSE optimization
- The scalar type of ```angular_(jacobian|hessian)_``` is changed to float
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/ndt_omp.h#L517
- A loop in the calculation of ```score_gradient``` is unrolled (this brings a large improvement on the single-thread registration speed, because a large portion of time is spent for this calculation)
https://github.com/koide3/pcl/blob/8b6a5aa079c76b939beacb161a81a54f712a1dc0/registration/include/pcl/registration/impl/ndt_omp.hpp#L394

koide3 commented:
> Is an update to non-special needed? Is there any reason we are not just removing original NDT and renaming ndt_omp to ndt besides maybe potential bugs?

Some concerns about replacing the original ndt with the new one are:
- The new code requires more memory during registration, because it allocates score, gradient, and hessian for each input point for thread safety
- The readability of ```updateDerivatives``` gets a bit worse in exchange for SSE optimization
- Does the policy of PCL allow to use OpenMP in a method whose name doesn't explicitly indicate its a multi-threaded algorithm? I thought PCL intentionally separate sinle-thread and multi-threaded algorithms like ```pcl::NormalEstimation``` and ```pcl::NormalEstimationOMP``` (I don't really know...)

kunaltyagi commented:
Regarding the last point:

OMP and normal implementations started diverging and it was decided to bring both under one umbrella and to use `#pragma omp if` to achieve that.

We'll be seeing some larger unification of API across PCL :crossed_fingers: Hopefully, all goes well

If possible something along the lines of PRs started by @shrijitsingh99 would be better :smile:

koide3 commented:
OK, the other concerns don't matter IMO. I think we can use less memory consuming routines when num_threads = 1. Should we append the new additions to this PR https://github.com/PointCloudLibrary/pcl/pull/4180 (or open another PR)?

koide3 commented:
OK, thanks! I'll push updates here.

koide3 commented:
I pushed multi-threaded NDT code. In addition to https://github.com/PointCloudLibrary/pcl/pull/4277#discussion_r457289896, I made two modifications:

-  ```trialValueSelectionMT()``` in ndt.hpp
Since the execution order of a multi-threaded loop can be changed, ```computeDerivatives()``` may return slightly different scores for the same input due to the floating point summation error. This can cause a condition where ```a_l == a_t``` but ```f_t != f_l``` (```f_*``` is scores, ```a_*``` are line search values) in ```trialValueSelectionMT()``` that results in yielding NaN at: https://github.com/PointCloudLibrary/pcl/blob/ff9adcf77109a1d510b51c393e8f56306123615a/registration/include/pcl/registration/impl/ndt.hpp#L481
To avoid this, I added some condition checks in ```trialValueSelectionMT()```: https://github.com/koide3/pcl/blob/dcb4be171d2e504c85c7f976496a7b719cc4ae6e/registration/include/pcl/registration/impl/ndt.hpp#L587

- test_ndt.cpp
NDT doesn't use kdtree, and the following test code doesn't make sense: https://github.com/PointCloudLibrary/pcl/blob/ff9adcf77109a1d510b51c393e8f56306123615a/test/registration/test_ndt.cpp#L75
I removed this test and added test for the consistency of single and multi-threaded results.

I'm testing the correctness of the updated NDT on KITTI. I'll report the test result once it's finished.

kunaltyagi commented:
Try not to force-push. Reviewers loose all context

koide3 commented:
Oh, sorry, I'll avoid force pushing

kunaltyagi commented on registration/include/pcl/registration/ndt.h:
```diff
@@ -216,6 +228,30 @@ namespace pcl
+        * \param nr_threads the number of hardware threads to use (0 sets the value back to automatic)
+        */
+      void setNumberOfThreads(unsigned int nr_threads = 0) {
+        if (nr_threads == 0) {
+        #ifdef _OPENMP
+          num_threads_ = omp_get_num_procs();
+        #else
+          num_threads_ = 1;
+        #endif
+        }
+        else {
+          num_threads_ = nr_threads;
+        }
```
```suggestion
#ifdef _OPENMP
        num_threads_ = nr_threads ? nr_threads : omp_get_num_procs();
#else
        PCL_DEBUG("OpenMP is not available. Keeping number of threads unchanged at 1");
        num_threads_ = 1;
#endif
```

koide3 replied:
I updated the code as suggested.

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -66,6 +66,9 @@ NormalDistributionsTransform<PointSource, PointTarget>::NormalDistributionsTrans
   transformation_epsilon_ = 0.1;
   max_iterations_ = 35;
+
+  setNumberOfThreads(0);
```
PCL defaults to 1 thread

koide3 replied:
Done

kunaltyagi commented on registration/include/pcl/registration/ndt.h:
```diff
@@ -455,19 +495,19 @@ namespace pcl
         * The precomputed angular derivatives for the jacobian of a transformation vector, Equation 6.19 [Magnusson 2009]. 
         */
-      Eigen::Matrix<double, 8, 4> angular_jacobian_;
+      Eigen::Matrix<float, 8, 4> angular_jacobian_;
```
Does the change from `double` to `float` provide computational benefit? SSE intrinsics for double do exist

koide3 replied:
When I wrote the code a few years ago, the float-version was significantly faster than the double-version on a bit old PC. But, I did bench-marking again today, and now the double-version was a bit faster than the float-version interestingly.

I guess this stems from AVX that has 256bit width and can process four doubles at a time. Indeed, I confirmed that the double-version got slower when AVX was disables.

I think AVX is very common now, and the use of float is not worthy. I replaced floats with doubles in the last commit.

(align 17k points 100 times)
```
Eigen::SimdInstructionSetsInUse(): AVX SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2
float :17153[msec]
double:16844.1[msec]

Eigen::SimdInstructionSetsInUse(): SSE, SSE2, SSE3, SSSE3, SSE4.1, SSE4.2
float :17613[msec]
double:18299.2[msec]
````

kunaltyagi commented on registration/include/pcl/registration/ndt.h:
```diff
@@ -288,6 +324,8 @@ namespace pcl
       double
       updateDerivatives (Eigen::Matrix<double, 6, 1> &score_gradient,
                          Eigen::Matrix<double, 6, 6> &hessian,
+                         const Eigen::Matrix<float, 4, 6>& point_gradient,
```
Can we prevent a complete breakage of API here? Is it possible to instead provide a migration path:
1. deprecate the `point_jacobian_` and `point_hessian_`
2. deprecate the original function declaration, make it check `num_threads_` and call the new implementation
3. default the new implementation

koide3 replied:
1. I reverted ```point_jacobian_``` and ```point_hessian_``` and added deprecation flags to them
2. I reverted their declaration and definitions

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -335,22 +389,26 @@ NormalDistributionsTransform<PointSource, PointTarget>::updateDerivatives (Eigen
+      // Update gradient, Equation 6.12 [Magnusson 2009]
+      x_trans_c_inv_hessian_ij.noalias() = x_trans_c_inv * point_hessian.block<4, 6>(i * 4, 0);
+
+      for (int j = 0; j < hessian.cols(); j++) {
```
Not `int` but `Eigen::Index`

koide3 replied:
Done

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -390,16 +476,36 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeHessian (Eigen::M
+  }
+
+  hessian.setZero();
+  for (std::size_t idx = 0; idx < num_threads_; idx++) {
```
`std::accumulate`?

koide3 replied:
Done.
I also updated https://github.com/koide3/pcl/blob/9c3fafe16d6e65addb64e5214f954659d1ef3cf5/registration/include/pcl/registration/impl/ndt.hpp#L240

kunaltyagi reviewed (commented):
You mentioned an unrolled loop, but I can't find it

koide3 commented:
> You mentioned an unrolled loop, but I can't find it

I meant the following lines. I moved the calculation of ```score_gradient``` outside of the loop and let Eigen do it as a fixed-size matrix multiplication. This ```updateDerivatives()``` was the biggest hot spot of the algorithm, and this unrolling had a certain influence on the entire processing speed.

https://github.com/PointCloudLibrary/pcl/blob/a90d982b4772ccf505d3365cd1fcf78e76f93b22/registration/include/pcl/registration/impl/ndt.hpp#L344
https://github.com/koide3/pcl/blob/0e0e77d6cfc1dd5e216fb9d74650ef82d33171ee/registration/include/pcl/registration/impl/ndt.hpp#L383

koide3 commented on registration/include/pcl/registration/ndt.h:
```diff
@@ -40,16 +40,28 @@
+   *       DIRECT1 uses the voxel, which the point fell into (up to 1 voxel per point). This makes registration extremely fast but a bit unstable in particular when the initial guess is not accurate.
+   *       DIRECT7 uses the voxel at the point and its facing voxels (up to 7 voxels per point). This strikes a good balance between registration stability and speed in a typical use case.
+   *       DIRECT27 uses the voxel at the point and its 3x3x3 neighboring voxels (up to 27 voxels per point).
+   *       KDTREE uses a kdtree search (radius = voxel resolution) to find neighbor voxels. With this method, the registration result will be identical to the result of the orginal NDT implementation.
```
I wanted to say that its result is the same as the result of the old PCL implementation (In the new code, DIRECT7 is the default method). I reworded the sentence as follows. 

```
   *       KDTREE uses a kdtree search (radius = voxel resolution) to find neighbor voxels. With this method, the registration result will be identical to the result of the implementation before PCL1.1.1.
```

kunaltyagi replied:
1.1.1? Perhaps you meant 1.11.1

koide3 replied:
Yeah, I corrected it

kunaltyagi commented:
Can we replace thread unsafe `computePointDerivatives` call with:
```cpp
void computePointDerivatives(const Eigen::Vector3d &x, bool compute_hessian)
{
  if (num_threads_ > 1) { PCL_WARN("This implementation is thread unsafe. Please use ..."); }
  computePointDerivatives(x, point_jacobian_, point_hessian_, compute_hessian);
}
```

This is better than a note which most ppl wouldn't read

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
````diff
@@ -242,35 +278,34 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeAngleDerivatives
 }
 
-
+// \note Thread-unsafe ```computePointDerivatives``` is deprecated
````
Can we replace thread unsafe `computePointDerivatives` call with:
```cpp
void computePointDerivatives(const Eigen::Vector3d &x, bool compute_hessian)
{
  if (num_threads_ > 1) { PCL_WARN("This implementation is thread unsafe. Please use ..."); }
  computePointDerivatives(x, point_jacobian_, point_hessian_, compute_hessian);
}
```

This is better than a note which most ppl wouldn't read. It'd also reduce code duplication

koide3 replied:
I replaced the thread unsafe implementations with calling thread safe ones

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
````diff
@@ -313,6 +348,48 @@ NormalDistributionsTransform<PointSource, PointTarget>::computePointDerivatives
+  }
+}
+
+// \note Thread-unsafe ```updateDerivatives``` is deprecated
````
If it's possible with `computePointDerivatives`, something similar here too

koide3 replied:
Done

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -242,83 +278,139 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeAngleDerivatives
-    const Eigen::Vector3d d = point_angular_hessian.block<3, 1>(6, 0);
-    const Eigen::Vector3d e = point_angular_hessian.block<3, 1>(9, 0);
-    const Eigen::Vector3d f = point_angular_hessian.block<3, 1>(12, 0);
+    Eigen::Vector4d a(0, point_angular_hessian[0], point_angular_hessian[1], 0.0f);
```
Can we make `a` through `f` const?

koide3 replied:
Done

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -242,83 +278,139 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeAngleDerivatives
 NormalDistributionsTransform<PointSource, PointTarget>::updateDerivatives (Eigen::Matrix<double, 6, 1> &score_gradient,
                                                                            Eigen::Matrix<double, 6, 6> &hessian,
+                                                                           const Eigen::Matrix<double, 4, 6> &point_jacobian,
+                                                                           const Eigen::Matrix<double, 24, 6> &point_hessian,
```
Might as well request x_trans and c_inv to be Vector4d.

PS: Please include a note that 4d is needed instead of 3d because of auto-vec

koide3 replied:
Done

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -242,83 +278,139 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeAngleDerivatives
-    point_hessian_.block<3, 1>(9, 5) = c;
-    point_hessian_.block<3, 1>(12, 5) = e;
-    point_hessian_.block<3, 1>(15, 5) = f;
+    point_hessian.block<4, 1>((9/3)*4, 3) = a;
```
Any reason for using (9/3)*4? Perhaps, modify the note in the line above to say:
```
... block matrix starting at (4i, j) ...
```
and update the indices to be of the same form, i.e. 4*3, 4*4, 4*5

koide3 replied:
I changed it to ```(3*4, 3)``` and added a note to indicate that it corresponds to the (3, 3) element of the equation in the paper

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -169,23 +167,46 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeDerivatives (Eige
 
   // Update gradient and hessian for each point, line 17 in Algorithm 2 [Magnusson 2009]
-  for (std::size_t idx = 0; idx < input_->size (); idx++)
+  #pragma omp parallel for num_threads(num_threads_) schedule(guided, 8)
```
Any reason for specifying the schedule?

koide3 replied:
I tested different scheduling methods and empirically determined that guided scheduling with chunk size = 8 is fastest in this case. I added a note to describe this.

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -169,23 +167,46 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeDerivatives (Eige
+
     // Find neighbors (Radius search has been experimentally faster than direct neighbor checking.
     std::vector<TargetGridLeafConstPtr> neighborhood;
     std::vector<float> distances;
```
Perhaps, move `distances` and `neighborhood` to outside the loop, with some heuristic size if possible
* clear them before using them
* declare them as privatefirst to OMP

koide3 replied:
Done

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -200,11 +221,26 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeDerivatives (Eige
+    const int thread_idx = 0;
+    #endif
+
+    scores[thread_idx] += score_pt;
```
Perhaps, we could use OMP's `reduction(+: score, score_gradient, hessian)`. They're supported since 2011's OpenMP 3.1

koide3 replied:
I tried to use ```reduction```, but it seems it has issues with Eigen matrices and yields corrupted values. I'm not really sure, but I guess there are some alignment problems related to Eigen's optimization.

kunaltyagi reviewed (commented):
Just a few points to reduce code

koide3 commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -169,23 +167,46 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeDerivatives (Eige
   // Update gradient and hessian for each point, line 17 in Algorithm 2 [Magnusson 2009]
-  for (std::size_t idx = 0; idx < input_->size (); idx++)
+  #pragma omp parallel for num_threads(num_threads_) schedule(guided, 8)
+  for (int idx = 0; idx < static_cast<int>(input_->size ()); idx++)
```
Done

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
````diff
@@ -233,90 +276,149 @@ NormalDistributionsTransform<PointSource, PointTarget>::computeAngleDerivatives
 }
 
-
+// \note Thread-unsafe ```computePointDerivatives``` is deprecated
````
These notes aren't needed anymore. The declaration is deprecated and the usage has a warning.

koide3 replied:
Done

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -361,45 +467,112 @@ NormalDistributionsTransform<PointSource, PointTarget>::updateDerivatives (Eigen
     const auto& x_trans_pt = trans_cloud[idx];
 
     // Find nieghbors (Radius search has been experimentally faster than direct neighbor checking.
     std::vector<TargetGridLeafConstPtr> neighborhood;
```
Please move these also out of the for loop

koide3 replied:
Done

kunaltyagi commented on test/registration/test_ndt.cpp:
```diff
@@ -46,52 +46,60 @@
-      tree_recip->setInputCloud (src);
-    reg.setSearchMethodSource (tree_recip, force_cache_reciprocal);
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+TEST (PCL, NormalDistributionsTransform_KDTREE)
```
This looks like a job for https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#value-parameterized-tests

koide3 replied:
Done

kunaltyagi commented on registration/include/pcl/registration/impl/ndt.hpp:
```diff
@@ -514,7 +691,7 @@ NormalDistributionsTransform<PointSource, PointTarget>::trialValueSelectionMT (d
   }
   // Case 3 in Trial Value Selection [More, Thuente 1994]
-  if (std::fabs (g_t) <= std::fabs (g_l))
+  if (std::fabs (g_t) <= std::fabs (g_l) && a_t != a_l)
```
This (and related) change seems to belong to a separate PR.

koide3 replied:
OK, I'll open another PR. Note that this change must be merged before this NDT_OMP PR because the multi-threaded NDT can cause segfault without this additional condition check.

kunaltyagi reviewed (commented):
Could you please verify that the deprecated functions aren't used anywhere inside PCL (except any code explicitly testing them)

koide3 commented:
All the deprecated ones are all protected members, and they are not used in anywhere other than NDT.

kunaltyagi commented on test/registration/test_ndt.cpp:
```diff
@@ -70,37 +74,22 @@ void test_ndt(NormalDistributionsTransform<PointXYZ, PointXYZ>& reg) {
+    NeighborSearchMethod::DIRECT7,
+    NeighborSearchMethod::DIRECT27
+  ),
+  [](const testing::TestParamInfo<NormalDistributionsTransformTest::ParamType>& info) -> std::string {
```
Would `const auto&` work? 😆

koide3 replied:
Done

kunaltyagi reviewed (commented):
LGTM. Minor comment only

koide3 commented:
Can I do rebase, squash, and force-push to resolve the conflict?

kunaltyagi commented:
There's a conflict due to the merge of formatting in registration module

koide3 commented:
I resolved the conflict.

kunaltyagi commented:
GCC 16.04 is failing because the it doesn't accept the fourth lambda

koide3 commented:
I remove the lambda in INSTANTIATE_TEST_SUITE_P

kunaltyagi commented:
LGTM

koide3 commented:
I resolved conflicts.

larshg commented:
Should this be merged 🚀 ?

kunaltyagi commented:
Needs 1 more review. Currently, requested from mvieth, but please feel free to review instead Lars :)
